common/txmgr/address_state.go

```diff
@@ -218,23 +274,196 @@ func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) MoveUn
 }
 
 // MoveTxToFatalError moves a transaction to the fatal error state.
+// It returns an error if there is no transaction with the given ID.
+// It returns an error if the transaction is not in an expected state.
 func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) MoveTxToFatalError(
 	txID int64, txError null.String,
 ) error {
+	as.Lock()
+	defer as.Unlock()
+
+	tx := as.allTxs[txID]
+	if tx == nil {
+		return fmt.Errorf("move_tx_to_fatal_error: no transaction with ID %d", txID)
+	}
+
+	as.moveTxToFatalError(tx, txError)
+
+	switch tx.State {
+	case TxUnstarted:
+		_ = as.unstartedTxs.RemoveTxByID(txID)
+	case TxInProgress:
+		as.inprogressTx = nil
+	case TxConfirmedMissingReceipt:
+		delete(as.confirmedMissingReceiptTxs, tx.ID)
+	default:
+		return fmt.Errorf("move_tx_to_fatal_error: transaction with ID %d is in an unexpected state: %s", txID, tx.State)
+	}
```

Comment:
Couldn't we also move a tx from Unconfirmed to FatalError?

Reply:
Oh not sure if we do that currently but if this is a method that generically moves a tx to `FatalError`, I was thinking we'd want it to be able to from any state. Otherwise we should maybe call out that this function wouldn't work for tx in Unconfirmed state?

Reply:
Sorry I forget sometimes that these methods also contain some validation and aren't purely updates. If these states should throw an error, I think you can leave it like it was before. At least going from Confirmed to FatalError sounds like it would be a bug we'd want to error on.